src/include/migraphx/op/reshape.hpp

```diff
@@ -46,14 +47,58 @@ struct reshape
     value attributes() const { return {{"require_std_shape", true}}; }
 
     std::string name() const { return "reshape"; }
-    shape compute_shape(std::vector<shape> inputs) const
+
+    shape dyn_compute_shape(shape s0) const
     {
-        check_shapes{inputs, *this}.has(1).standard();
+        auto dyn_dims      = s0.dyn_dims();
+        auto num_not_fixed = std::count_if(
+            dyn_dims.cbegin(), dyn_dims.cend(), [](auto dd) { return not dd.is_fixed(); });
+        if(num_not_fixed != 1)
+        {
+            MIGRAPHX_THROW("Reshape: Only supports one non-fixed dynamic_dimension");
+        }
+        std::size_t not_fixed_index = 0;
+        // track number of fixed elements in input and output
+        std::size_t num_dims_ele = 1;
+        std::size_t num_dd_ele   = 1;
+        for(std::size_t i = 0; i < dyn_dims.size(); ++i)
+        {
+            if(dyn_dims[i].is_fixed())
+            {
+                num_dims_ele *= dims[i];
+                num_dd_ele *= dyn_dims[i].min;
+            }
+            else
+            {
+                not_fixed_index = i;
+            }
+        }
+        if(num_dims_ele != num_dd_ele)
+        {
+            MIGRAPHX_THROW("Reshape: Number of fixed elements must match. Input: " +
+                           std::to_string(num_dd_ele) + " Output: " + std::to_string(num_dims_ele));
+        }
+        if(dims[not_fixed_index] != 0 and dims[not_fixed_index] != -1)
+        {
+            MIGRAPHX_THROW("Reshape: Non-fixed dynamic_dimension doesn't match with 0 or -1 "
+                           "output dimension");
+        }
+        // construct output dynamic shape from dims attribute
+        std::vector<shape::dynamic_dimension> output_dyn_dims(dims.size());
+        std::transform(dims.cbegin(), dims.cend(), output_dyn_dims.begin(), [](auto in_d) {
+            std::size_t d = in_d;
+            return shape::dynamic_dimension{d, d};
+        });
+        output_dyn_dims[not_fixed_index] = dyn_dims[not_fixed_index];
```

Comment:
This should be done in the `std::transform` as well:

```cpp
std::transform(dims.cbegin(), dims.cend(), dyn_dims.begin(), std::back_inserter(output_dyn_dims), [](std::size_t dim, auto dyn_dim) {
    if (dyn_dim.is_fixed())
        return dyn_dim;
    return shape::dynamic_dimension{dim, dim};
});
```

Reply:
I've been wondering about `transform` and `back_inserter` since `transform` doesn't guarantee in-order traversal of the iterators. It works for our setup, but might not be up to the standard spec?

Reply:
The `back_inserter` is an `OutputIterator` so it only allows traversing it one way. Although it could traverse the inputs in a different order, but this is not a problem because we dont have any side effects in the binary_op we pass.